--- varnish-2.0.6/configure.ac~	2013-12-13 12:30:34.000000000 +0200
+++ varnish-2.0.6/configure.ac	2013-12-13 12:31:38.099161265 +0200
@@ -375,22 +375,23 @@
 # Use jemalloc on Linux
 JEMALLOC_SUBDIR=
 JEMALLOC_LDADD=
-AC_ARG_ENABLE(jemalloc,
-AS_HELP_STRING([--disable-jemalloc],[do not use jemalloc (default is yes on Linux, no everywhere else)]),
-[if "x$enableval" = "xyes"; then
-	JEMALLOC_SUBDIR=libjemalloc
-	JEMALLOC_LDADD='$(top_builddir)/lib/libjemalloc/libjemalloc_mt.la'
-fi],
-[case $target in #(
-*-*-linux*)
-	JEMALLOC_SUBDIR=libjemalloc
-	JEMALLOC_LDADD='$(top_builddir)/lib/libjemalloc/libjemalloc_mt.la'
-	;; #(
-*)
-	true
-	;;
-esac])
+AC_ARG_WITH([jemalloc],
+            [AS_HELP_STRING([--with-jemalloc],
+              [use jemalloc memory allocator.  Default is yes on Linux,  no elsewhere])],
+            [],
+            [with_jemalloc=check])
 
+case $target in
+    *-*-linux*)
+        if test "x$with_jemalloc" != xno; then
+            AC_CHECK_LIB([jemalloc], [malloc],
+                  [JEMALLOC_LDADD="-ljemalloc"],
+                  [AC_MSG_NOTICE([No system jemalloc found, using bundled version])
+                   JEMALLOC_SUBDIR=libjemalloc
+                   JEMALLOC_LDADD='$(top_builddir)/lib/libjemalloc/libjemalloc_mt.la'])
+	fi
+	;;
+esac
 AC_SUBST(JEMALLOC_SUBDIR)
 AC_SUBST(JEMALLOC_LDADD)
 
