--- varnish-2.1.4/lib/libvarnish/Makefile.am~	2010-12-03 14:20:32.000000000 +0200
+++ varnish-2.1.4/lib/libvarnish/Makefile.am	2010-12-03 14:21:04.038837288 +0200
@@ -31,7 +31,8 @@
 	vss.c \
 	vtmpfile.c
 
-libvarnish_la_LIBADD = ${RT_LIBS} ${NET_LIBS} ${LIBM} @PCRE_LIBS@
+libvarnish_la_LIBADD = ${RT_LIBS} ${NET_LIBS} ${LIBM} @PCRE_LIBS@ \
+	$(top_builddir)/lib/libvarnishcompat/libvarnishcompat.la
 
 DISTCLEANFILES = svn_version.c
 svn_version.c: FORCE
