--- varnish-2.1.4/lib/libvarnish/Makefile.am~	2010-12-03 14:20:32.000000000 +0200
+++ varnish-2.1.4/lib/libvarnish/Makefile.am	2010-12-03 14:21:04.038837288 +0200
@@ -31,7 +31,8 @@
 	vss.c \
 	vtmpfile.c
 
-libvarnish_la_LIBADD = ${RT_LIBS} ${NET_LIBS} ${LIBM} @PCRE_LIBS@
+libvarnish_la_LIBADD = ${RT_LIBS} ${NET_LIBS} ${LIBM} @PCRE_LIBS@ \
+	$(top_builddir)/lib/libvarnishcompat/libvarnishcompat.la
 
 DISTCLEANFILES = svn_version.c
 svn_version.c: FORCE
--- varnish-2.1.5/lib/libvcl/Makefile.am~	2011-08-14 01:34:34.000000000 +0300
+++ varnish-2.1.5/lib/libvcl/Makefile.am	2011-08-14 01:34:37.768099966 +0300
@@ -27,6 +27,9 @@
 	vcc_var.c \
 	vcc_xref.c
 
+libvcl_la_LIBADD = \
+	$(top_builddir)/lib/libvarnish/libvarnish.la
+
 EXTRA_DIST = \
 	vcc_gen_fixed_token.tcl
 
